### DISCUSSION TITLE
cgmDat

### DISCUSSION THREAD

**Discussion by jjburton** [jjburton/cgmTools#296] [General]
Create a source data class for various data types to use.

- [x] save/load
- [x] read/write
- [x] decoding

Rules
- Dict keys must be strings and not ints or other data types
-

**Comment by jjburton** [comment]
### Week of 11.4

```python
import cgm.core.lib.distance_utils as DIST
help(DIST.get_pos_by_linearPct)


import cgm.core.cgm_Dat as CGMDAT
reload(CGMDAT)

mDat = CGMDAT.data()
mDat.dataformat = 'json'
mDat.write()

mDat.read()

mDat.log_self()
mDat.log_dat()

CGMDAT.decodeDat(mDat)
mDat.decodeDat()

mDat.dat['test'] = range(3)
mDat.dat['dictTest'] = {'cat':1,'dog':2}
mDat.dat['dictNext'] = {'d1':{'cat':1,'dog':2},'list':[1,2,5]}
```

Questions to answer
- Where should a local scene's block dat save?

content/cgmDat/mrs/
file/cgmDat/mrs/blocktype/blockNamestring

**Comment by jjburton** [comment]
### MRS Dat
- blockDat | individual Block
- shapeDat | specific shape dat
- mrsConfig | collection of block dats, needs parent/children wiring, needs to be ordered

**Comment by jjburton** [comment]
### BlockDat
- [x] initial class
- [x] get block dat
- [x] store block dat
- [x] load block dat 
- [x] save block dat
- [x] load block dat from file


```python
import cgm.core.mrs.MRSDat as MRSDAT
reload(MRSDAT)

mDat = MRSDAT.BlockDat('master_masterBlock')
mDat.mBlock.atUtils('blockDat_get')
mDat.get()
mDat.set()
mDat.log_self()
mDat.log_dat()

mDat.save()
mDat.load()

mDat.write()
mDat.read()

reload(mDat.mBlock.p_blockModule)
MRSDAT.blockDat_get(mDat.mBlock)
MRSDAT.blockDat_load(mDat.mBlock,mDat.dat)
MRSDAT.blockDat_load(mDat.mBlock,mDat.dat,currentOnly =True)

```

**Comment by jjburton** [comment]
How do we wanna store these.
- What do we call them - Presets?Prefig? Need more than one  block support 

-Prop
-- Weapon
---- sword
-------handle

cgmDat.mrs.shapeDat.leg.tigerFront

**Reply by benngarnish** [reply]
How about templates?

**Reply by jjburton** [reply]
I use that already with the end files. Worried using that again might be confusing.

**Reply by benngarnish** [reply]
Its the same thing almost tho right?
We are creating templates either by saving data or by creating manually as a block and saving a maya file.
**Templates** = The act of saving blocks, whether be in a maya file or data file that represents your character/assets or piece of a final rig file.

**Reply by jjburton** [reply]
No a template is a full buildable asset. These won't all have masters or full setups unless explicitly designed as such - biped human male for example. I want to be able to have an arm with fingers for example.

**Reply by benngarnish** [reply]
Components?
Or
Configurations?

**Reply by jjburton** [reply]
maybe? Nothing jumping at me yet.

**Comment by jjburton** [comment]
### Week of 11.8
- [ ] save to repos
- [ ] save to custom path
- [ ] build from blockDat
- [ ]

**Comment by jjburton** [comment]
### No more bakeScale
Work through getting rid of the idea of bakeScale

**Reply by benngarnish** [reply]
This has me smiling.

**Comment by jjburton** [comment]
### Handle
Cube
- [x] define state
- [x] form state
- [x] prerig
- [ ] master
- [ ] scale pivot
- [ ] aim

Shaper list stuff
- [ ] define state
- [ ] form state
- [ ] prerig


Fix
- [ ] make the following use joint radius - scalePivot helper, aim

Notes
- Removed blockProfile from the name list. Feels too forced to make that necessary and will really increase the block names
- Explored getting rid of bake scale and the simplest solution feels like just having the rebuild sort the scaling change

**Comment by jjburton** [comment]
### UI
- [x] Get Dat
- [x] Load Dat
- [x] Save Dat
- [x] Print dat

**Comment by jjburton** [comment]
### Start dir
- workspace | start at the workspace
- project | cgm project path
- file | put it where the file is
- dev | for MRS saving settings


There was an issue using the file mode when we don't have saved file.

**Comment by jjburton** [comment]
### Block dat initial functions
- [x] Get data from selected
- [x] Load data to ui
- [x] save data from ui
- [x] Create master
- [x] Create handle
- [x] Create Limb
- [ ] Create Head
- [ ] Load data to master
- [ ] Load data to limb
- [ ] Load data to segment
- [ ] Load data to head

**Comment by jjburton** [comment]
### Issues 11.19
- [x] Limb | joint vis radius shouldn't be under no trans but trans define
- [x] Building limb from dat isn't getting foot shapes properly
- [x] Shape dat isn't grabbing the end shapes
